rflx/pyrflx/typevalue.py

```diff
@@ -766,6 +794,109 @@ def _preset_fields(self, fld: str) -> None:
             self._last_field = nxt
             nxt = self._next_field(nxt)
 
+    def set_checksum_function(self, checksum_field_name: str, checksum_method: Callable) -> None:
+        if checksum_field_name not in self.fields:
+            raise KeyError(f"Field {checksum_field_name} is not defined")
+        for field_name, checksum in self._checksums.items():
+            if field_name == checksum_field_name:
+                checksum.set_checksum_function(checksum_method)
+                return
+        raise KeyError(f"Field {checksum_field_name} has not been defined as a checksum field")
+
+    def _is_checksum_settable(self, checksum_aspect: "MessageValue.Checksum") -> bool:
+        def valid_path(value_range: ValueRange) -> bool:
+            expr: Dict[Expr, str] = dict.fromkeys([value_range.lower, value_range.upper])
+
+            for k in expr:
+                if isinstance(k, Sub):
+                    assert isinstance(k.left, (First, Last))
+                    expr[k] = str(k.left.prefix)
+                elif isinstance(k, Add):
+                    for t in k.terms:
+                        if isinstance(t, (First, Last)):
+                            expr[k] = str(t.prefix)
+                else:
+                    assert isinstance(k, (First, Last))
+                    expr[k] = str(k.prefix)
+
+            field = expr.get(value_range.lower)
+            assert isinstance(field, str)
+            upper_field_name = expr.get(value_range.upper)
+            if upper_field_name == "Message":
+                upper_field_name = "Final"
+            while field != upper_field_name:
+                field = self._next_field(field)
+                if field == "Final":
+                    continue
+                if field == "" or not self._fields[field].set:
+                    break
+            else:
+                return True
+
+            return False
+
+        for expr in checksum_aspect.parameters:
+            expr.evaluated_expression = self.__simplified(copy.copy(expr.expression))
```

Comment:
`expr` sounds like the elements would be of type `Expr`, but that is not the case.

Reply:
Renamed in c238ac0